File: CoreVia-Android/app/src/main/java/life/corevia/app/data/model/ChatModels.kt
package life.corevia.app.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class ChatConversation(
    val id: String = "",
    @SerialName("user_id") val usersId: String = "",
    @SerialName("user_name") val userName: String = "",
    @SerialName("user_profile_image") val userProfileImage: String? = null,
    @SerialName("user_type") val userType: String = "client",
    @SerialName("last_message") val lastMessage: String? = null,
    @SerialName("last_message_time") val lastMessageTime: String? = null,
    @SerialName("unread_count") val unreadCount: Int = 0,
    @SerialName("is_trainer") val isTrainer: Boolean = false
)

@Serializable
data class ChatMessage(
    val id: String = "",
    @SerialName("sender_id") val senderId: String = "",
    @SerialName("receiver_id") val receiverId: String = "",
    @SerialName("sender_name") val senderName: String = "",
    @SerialName("sender_profile_image") val senderProfileImage: String? = null,
    val message: String = "",
    @SerialName("is_read") val isRead: Boolean = false,
    @SerialName("created_at") val createdAt: String = ""
)

@Serializable
data class ChatMessageCreate(
    @SerialName("receiver_id") val receiverId: String,
    val message: String
)

@Serializable
data class MessageLimitResponse(
    val limit: Int = 50,
    val used: Int = 0,
    val remaining: Int = 50
)

File: CoreVia/CoreVia/Models/ChatModels.swift
//
//  ChatModels.swift
//  CoreVia
//

import Foundation
import os.log

// MARK: - Chat Message Response
struct ChatMessageResponse: Codable, Identifiable {
    let id: String
    let senderId: String
    let receiverId: String
    let senderName: String
    let senderProfileImage: String?
    let message: String
    let isRead: Bool
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id, message
        case senderId = "sender_id"
        case receiverId = "receiver_id"
        case senderName = "sender_name"
        case senderProfileImage = "sender_profile_image"
        case isRead = "is_read"
        case createdAt = "created_at"
    }
}

// MARK: - Chat Message Create
struct ChatMessageCreate: Codable {
    let receiverId: String
    let message: String

    enum CodingKeys: String, CodingKey {
        case receiverId = "receiver_id"
        case message
    }
}

// MARK: - Chat Conversation
struct ChatConversation: Codable, Identifiable {
    var id: String { userId }
    let userId: String
    let userName: String
    let userProfileImage: String?
    let lastMessage: String
    let lastMessageTime: Date
    let unreadCount: Int

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case userName = "user_name"
        case userProfileImage = "user_profile_image"
        case lastMessage = "last_message"
        case lastMessageTime = "last_message_time"
        case unreadCount = "unread_count"
    }
}

// MARK: - Message Limit
struct MessageLimitResponse: Codable {
    let dailyLimit: Int
    let usedToday: Int
    let remaining: Int

    enum CodingKeys: String, CodingKey {
        case dailyLimit = "daily_limit"
        case usedToday = "used_today"
        case remaining
    }
}

// MARK: - Chat Manager
class ChatManager: ObservableObject {
    static let shared = ChatManager()

    @Published var conversations: [ChatConversation] = []
    @Published var messages: [ChatMessageResponse] = []
    @Published var messageLimit: MessageLimitResponse?
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let api = APIService.shared
    private init() {}

    @MainActor
    func fetchConversations() async {
        isLoading = true
        errorMessage = nil
        do {
            let result: [ChatConversation] = try await api.request(
                endpoint: "/api/v1/chat/conversations"
            )
            conversations = result
        } catch {
            AppLogger.network.error("Conversations fetch xetasi: \(error.localizedDescription)")
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    @MainActor
    func fetchMessages(userId: String) async {
        isLoading = true
        do {
            let result: [ChatMessageResponse] = try await api.request(
                endpoint: "/api/v1/chat/history/\(userId)"
            )
            messages = result
        } catch {
            AppLogger.network.error("Messages fetch xetasi: \(error.localizedDescription)")
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    @MainActor
    func sendMessage(receiverId: String, message: String) async -> Bool {
        errorMessage = nil
        do {
            let body = ChatMessageCreate(receiverId: receiverId, message: message)
            let result: ChatMessageResponse = try await api.request(
                endpoint: "/api/v1/chat/send",
                method: "POST",
                body: body
            )
            messages.append(result)
            await fetchMessageLimit()
            return true
        } catch let error as APIError {
            errorMessage = error.errorDescription
            return false
        } catch {
            AppLogger.network.error("Send message xetasi: \(error.localizedDescription)")
            errorMessage = error.localizedDescription
            return false
        }
    }

    @MainActor
    func fetchMessageLimit() async {
        do {
            let result: MessageLimitResponse = try await api.request(
                endpoint: "/api/v1/chat/limit"
            )
            messageLimit = result
        } catch {
            AppLogger.network.error("Message limit fetch xetasi: \(error.localizedDescription)")
        }
    }
}
